skunkworks_crow/src/main/java/org/odk/share/views/ui/bluetooth/BtSenderActivity.java

```diff
@@ -73,6 +75,11 @@ protected void onCreate(Bundle savedInstanceState) {
         setTitle(getString(R.string.send_instance_title));
         setSupportActionBar(toolbar);
 
+        //checking the location permission for bluetooth for android 9.
```

Comment:
I think it would be better to ask for this permission before opening the activity(When bluetooth option is choosen). Because both the permission asking dialog and scanning dialog appears at the same time .

Reply:
Can we grant this permission when we are entering the main page? If we choose bluetooth then popup the permission dialog, that page still has the method choosing dialog appears at the same time.

Reply:
If we ask for location permission on start, then what if the user never chooses `Bluetooth` as the transfer mode then it would not be of no use. 
I'm suggesting to open a dialog asking for user permission when user taps on `Bluetooth` as the transfer mode(Close the dialog asking for the mode of transfer then only show the dialog for the permission) and if user allows then send to BluetoothReceiver/Sender Screen otherwise again show the dialog saying that `Location permission is needed for Bluetooth transfer` with two button options `Cancel`(negative) and `Access permission`(positive). `Access permission` will ask user again for the permission and if he has selected the checkbox `Never ask again` when permission was asked then you won't be able to request for permission again, in that case, you have to navigate user to our apps' settings so that he can manually go and give permission to us.
I know its pretty long process and but I think this is the way to tackle this issue.

Reply:
This is how generally other apps tackle this, in case you've some other option in mind please let me know.

Reply:
Cool, I'll follow your suggestion. But can I use some run-time android permission library? So that I can remove the `PermissionUtil.java` and make the code more flexible, something like https://github.com/Karumi/Dexter, it can also help me to build a dialog.

Reply:
Sure, It totally depends on you what you want to use, and but don't forget to add it in our `Open Source Licenses` section, its always to good to give the **credits**. And yes if you're doing it for Location permission then do it for every dangerous permission like `read and write storage`. Or you can open a different PR for that.